Assets/MRTK/Core/Utilities/Async/Internal/AsyncCoroutineRunner.cs

```diff
@@ -56,30 +56,19 @@ internal static AsyncCoroutineRunner Instance
 
                 if (instance == null)
                 {
-                    var instanceGameObject = GameObject.Find("AsyncCoroutineRunner");
+                    instance = FindObjectOfType<AsyncCoroutineRunner>(true);
 
-                    if (instanceGameObject != null)
+                    if (instance != null)
                     {
-                        instance = instanceGameObject.GetComponent<AsyncCoroutineRunner>();
-
-                        if (instance == null)
-                        {
-                            Debug.Log("[AsyncCoroutineRunner] Found GameObject but didn't have component");
-
-                            if (Application.isPlaying)
-                            {
-                                Destroy(instanceGameObject);
-                            }
-                            else
-                            {
-                                DestroyImmediate(instanceGameObject);
-                            }
-                        }
+                        Debug.LogError($"GameObject {instance.gameObject.name} has an inactive AsyncCoroutineRunner attached and " +
+                        "there is no active AsyncCoroutineRunner in the scene. There must be an active AsyncCoroutineRunner attached to a GameObject at the root of the scene.");
+                        return instance;
                     }
                 }
 
                 if (instance == null)
                 {
+                    Debug.LogWarning("There is no AsyncCoroutineRunner in the scene. Adding a GameObject with AsyncCoroutineRunner attached at the root of the scene.");
```

Comment:
Does this need to be a warning if the toolkit is adding it on the user's behalf?

Reply:
@davidkline-ms the Instance creation in AwaiterExtensions.cs only causes AsyncCoroutineRunner to be created when you're using AwaiterExtensions. It is still possible to call AsyncCoroutineRunner.Post directly without using AwaiterExtensions.

Reply:
> It is still possible to call AsyncCoroutineRunner.Post directly without using AwaiterExtensions.

@jbienzms Can you expand on this? I guess we could write additional MRTK code, but `Post` itself is `internal`. I'm fine leaving `internal` implementations in a specific known state.

Reply:
Well, I may be wrong about this. But the way I found this bug was that [WindowsSceneUnderstandingObserver](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/main/Assets/MRTK/Providers/Experimental/WindowsSceneUnderstanding/WindowsSceneUnderstandingObserver.cs) calls `RunObserverAsync` on line 157. This in turn calls `WaitForUpdate` on line 663. This eventually ends up in `AsyncCoroutineRunner.Post`. What I don't know is what happens between `CustomYieldInstruction.WaitForUpdate` and `AsyncCoroutineRunner.Post`, because I can't seem to find the code for `CustomYieldInstruction` using Github search in the MRTK repo. This may mean that `CustomYieldInstruction` is in a binary library. 

Please verify that the path between `CustomYieldInstruction.WaitForUpdate` and `AsyncCoroutineRunner.Post` will guarantee that the `AsyncCoroutineRunner` instance gets created.

Reply:
Under this PR `WaitForUpdate` will properly trigger a call to get the instance (and warn user if it fails) as `WaitForUpdate` eventually calls `RunOnUnityScheduler` where the new call to get the instance is added. You can perform a find all references in VS on `WaitForUpdate` and trace the result in `AwaiterExtensions` if you want to see the call hierarchy. So `AsyncCoroutineRunner.Post` does not guarantee the existence of the instance but the only caller to `AsyncCoroutineRunner.Post`, namely `RunOnUnityScheduler`, does.

Reply:
Yep! From `await new WaitForUpdate()`, we get into AwaiterExtensions at

https://github.com/microsoft/MixedRealityToolkit-Unity/blob/ae93bae2bf9ac6cbb0cc70678e41a37e4dc05b92/Assets/MRTK/Core/Utilities/Async/AwaiterExtensions.cs#L49-L52

then https://github.com/microsoft/MixedRealityToolkit-Unity/blob/ae93bae2bf9ac6cbb0cc70678e41a37e4dc05b92/Assets/MRTK/Core/Utilities/Async/AwaiterExtensions.cs#L124-L130

which calls RunOnUnityScheduler, which is the method being updated in this PR